kong/plugins/oauth2/secret.lua

```diff
@@ -0,0 +1,159 @@
+local utils = require "kong.tools.utils"
+
+
+local type = type
+local find = string.find
+local pcall = pcall
+local remove = table.remove
+local concat = table.concat
+local assert = assert
+
+
+--local CORES
+--do
+--  local infos = utils.get_system_infos()
+--  if type(infos) == "table" then
+--    CORES = infos.cores
+--  end
+--  if not CORES then
+--    CORES = ngx.worker.count() or 1
+--  end
+--end
+
+
+local PREFIX = nil -- currently chosen algorithm
+
+
+local ARGON2
+local ARGON2_ID = "$argon2"
+do
+  local ARGON2_PREFIX
+  local ok, crypt = pcall(function()
+    local argon2 = require "argon2"
+
+    -- argon2 settings
+    local ARGON2_VARIANT     = argon2.variants.argon2_id
+    local ARGON2_PARALLELISM = 1 --CORES
+    local ARGON2_T_COST      = 1
+    local ARGON2_M_COST      = 4096
+    local ARGON2_HASH_LEN    = 32
+    local ARGON2_SALT_LEN    = 16
+
+    local ARGON2_OPTIONS = {
+      variant     = ARGON2_VARIANT,
+      parallelism = ARGON2_PARALLELISM,
+      hash_len    = ARGON2_HASH_LEN,
+      t_cost      = ARGON2_T_COST,
+      m_cost      = ARGON2_M_COST,
+    }
+    do
+      local hash = argon2.hash_encoded("", utils.get_rand_bytes(ARGON2_SALT_LEN), ARGON2_OPTIONS)
+      local parts = utils.split(hash, "$")
+      remove(parts)
+      remove(parts)
+      ARGON2_PREFIX = concat(parts, "$")
+    end
+
+    local secret = {}
+
+    function secret.hash(secret)
+      return argon2.hash_encoded(secret, utils.get_rand_bytes(ARGON2_SALT_LEN), ARGON2_OPTIONS)
+    end
+
+    function secret.verify(secret, hash)
+      return argon2.verify(hash, secret)
+    end
+
+    return secret
+  end)
+
+  if ok then
+    ARGON2 = crypt
+    PREFIX = PREFIX or ARGON2_PREFIX
+  end
+end
+
+
+local BCRYPT
+local BCRYPT_ID = "$2"
+do
+  local BCRYPT_PREFIX
+  local ok, crypt = pcall(function()
```

Comment:
Same as before: wrap only the `require` with `pcall`.

Reply:
I am not sure what is the benefit (only seeing it more flaky). These only run once on first require.